## Introduction to the Machine Learning Workflow
Dataquest wants to figure out whether or not a new learner will complete a specific course. We are confident that the course has a good balance that allows a learner to understand any concept by applying it. Unfortunately, that doesn't always imply that everyone will complete a course.

So, how could we predict that a new learner will complete the course?

Let's say we've collected data for every current learner on the website. We know:

- How many lessons they've completed previously on the website.
- How many of the other courses they've completed.
- How many exercises they've successfully passed.
- How many hours they've spent on each individual lesson.
- Whether they've already completed the course

Based on the data, we already know whether an existing learner has completed the course. Are there relevant details in the rest of the data that could help us find a pattern that relates a learner's actions on the website to the likelihood of their completing the course?

Maybe, upon exploring the data, we find that 90% of learners who completed the course:

- Completed more than L lessons previously.
- Spent more than H hours, on average, per lesson.
- Passed more than E exercises.
- Given these patterns, can we say that a new learner who satisfies the above criteria will complete that course also?

Well, we can't guarantee it. But, the likelihood of that happening might be high if the new learner satisfies the above rules. What if we learn something new from the data that changes that likelihood? It could get lower or higher.

We can't always identify these patterns one at a time, especially when we have lots of data. This is exactly where machine learning can help.

Instead of making decisions based on patterns identified from information we already have, we can use machine learning to do it for us.

The small set of rules or criteria we identified above is essentially what a machine learning model does without us explicitly programming those rules. We train the model so it learns to identify those patterns from the data on its own. We then use the model to predict something given a new, unseen,

input. That "something" might seem a bit ambiguous, but what we want to predict depends on the intended outcome or task. In this and the next few courses, we will cover such tasks and learn machine learning models that can help us with those tasks.

What we learned above gives us a quick introduction to the machine learning workflow:

- Data Collection.
- Data Exploration and Wrangling.
- Data Preparation (Feature Engineering).
- Building and training a model.
- Evaluating the model performance
- Fine-tuning the model.
- Evaluating the model performance.

<img src='ML_workflow.svg' width=500 height=500>

That might seem like a lot of steps and terminology! Some things might not be explicitly clear right away, but don't worry. Throughout this course, we will go through each step of the above workflow in more detail.
For the rest of this lesson, we will use a popular, python-based machine learning library called scikit-learn and learn more about this workflow. We'll also train a simple machine learning model to predict if a patient has breast cancer.

Please note that this lesson will focus on helping us familiarize with the workflow. Any concepts or terminology that might come across as confusing or challenging will be covered in varying depths across the remaining lessons and courses

## Data Collection and Exploration
We'll build a model that predicts whether a patient has breast cancer. We'll start with the data.
While it's possible, it's not always feasible to collect our own data. Fortunately, there are a lot of publicly available datasets that we can freely access and work with, depending on what kind of problem we want to solve.

We'll work with the Breast Cancer Wisconsin (Diagnostic) Dataset https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic). The scikit-learn library stores several datasets, including this breast cancer dataset, and makes it convenient for us to load and work with them.

On the first screen, we looked at what kind of data could be collected by a company like Dataquest:

On the first screen, we looked at what kind of data could be collected by a company like Dataquest:

- How many lessons students have completed on the website.
- How many courses they've completed so far.
- How many exercises they've successfully passed.
- How many hours they've spent on each lesson.
Each of the above is called a **feature**. They all describe or are a property of our data. When we work with tabular data, each column name corresponds to a feature -- except one:of our data. When we work with tabular data, each column name corresponds to a feature -- except one:

Whether or not a learner has already completed the course.
The column corresponding to the above is called the **target variable** because that's what we want our model to predict. That's our target.

Every row for the above would contain information related to an individual learner. Each of these rows is called an **observation or a feature vector**. It's an n-dimensional vector of features

Let's look at the dataset we'll be working with. The link to our dataset provides us with the following information:

- There are a total of 30 attributes, or 30 features.
- There are a total of 569 instances or observations.
- There are two classes in the target variable:
- WDBC-Malignant
- WDBC-Benign
The above two classes, or labels, tell us whether a patient has a benign or a malignant tumor. For every observation, there's a corresponding class or label in our data.

Let's load in our dataset using scikit-learn's load_breast_cancer function https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer. The function will return an object with several attributes, a couple of which we will utilize:

- data stores the data points.
- target stores the value 0 if the tumor is benign or 1 if the tumor is malignant

In [10]:
#import necessary libraries

from sklearn.datasets import load_breast_cancer
#call the imported function
load_breast_cancer()



{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [11]:
#pass the as_frame=True parameter to the function
cancer_data=load_breast_cancer(as_frame=True)

#assign the data attribute to cancer_df
cancer_df=cancer_data.data
cancer_df['target']=cancer_data.target

In [12]:
#print the first five rows of the dataset
cancer_df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [13]:
cancer_df['target'].unique()

array([0, 1])

In [14]:
#print the shape of the data
cancer_df.shape

(569, 31)

In [15]:
#mssing values in each column
cancer_df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64

In [9]:
cancer_df['target'].value_counts()

1    357
0    212
Name: target, dtype: int64

In [17]:
#display the summary statistics of the data
cancer_df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      target  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

## Supervised Machine Learning: Classification
On the previous screen, we identified that our data has no missing values. We can, for now, work with the assumption that our data is clean and doesn't require further wrangling. That's always a dangerous assumption to make, but, for this lesson, our focus is on exploring the machine learning workflow from a broader perspective.

Our next step is to prepare our data before we input it to our model. However, we haven't really discussed what this model is.

Let's say we don't know what a giraffe looks like. Someone shows us a photo of a giraffe and we are able to identify some features that we think are unique to a giraffe. The next time we see a photo of a giraffe, even if it's a completely different one than before, we are highly likely to know that it's a giraffe. Our innate pattern-matching abilities could fail us, and we might confuse it for an ostrich just because of the long neck. But that's unlikely.

<img src='Giraffe.png' width=600 height=600>

We were supplied with a label, along with a set of features, and our brains learned to associate that label with those features. The next time we saw a similar set of features, we were able to predict the correct label. That's the intuition behind the subset of machine learning we'll learn about in this course--**supervised machine learning**

It's supervised because our model learns from existing data and the corresponding labels. There are usually two types of labels that we encounter:

- Continuous labels.
For example, we could be working with a dataset that contains features that describe different types of cars, and the labels could be the price of those cars. Our model would then learn to predict the price of a car, given those features as input.
- Categorical labels.
Our breast cancer dataset has only two labels--benign and malignant, or 0 and 1. Each observation is categorized by its own label or class.

If the label or target we want to predict is a categorical value, we call it a **classification task**. The model, a classifier, will try to classify a given set of inputs into a category. There are other types of machine learning models that can be used for different kinds of tasks. We'll learn about those later.

-So, what exactly does the model learn from the data?

This is dependent on the machine learning algorithm we end up using for our model. Throughout this course, we will cover several algorithms that will answer that question for us. But, intuitively, think of our dataset as a bunch of points on a chart:

<img src='chart.svg' width=500 height=500>

The points on the charts above are our features in a two-dimensional space. This space is called a **feature space** and the colors represent the labels. The line that we see is a **decision boundary** and it divides the feature space into two. Points on one side of the boundary belong to one class, and points on the other side of the boundary belong to the other class.

That decision boundary is our **classifier**. Since we only have two labels, it's called a **binary classifier**. If we had more labels, it would be a multi-class classifier.

<img src='multi_classifier.svg' width=500 height=500>

If we had more features, we would have an n-dimensional space and the decision boundary would be a hyperplane instead of a line.

But where does that line come from?

We know that a line or a curve can be defined using a parametric equation. That's what the model tries to learn - the parameters of that equation. The model estimates those parameters from the data, and those estimated parameters define the decision boundary.

If we had more features, we would have an n-dimensional space and the decision boundary would be a hyperplane instead of a line.

- But where does that line come from?

We know that a line or a curve can be defined using a parametric equation. That's what the model tries to learn - **the parameters** of that equation. The model estimates those parameters from the data, and those estimated parameters define the decision boundary.

Now that we have a better idea of what kind of machine learning task we're carrying out, let's prepare our data so that we can then train a model on it

## · Data Preparation
In the previous screens, we came across a couple of concepts:

- Training a model on our data.
- Using the model to make a prediction on unseen data.
The data that we use to train our model is called **training data**, a **training set** or a **training dataset**
The unseen data, as the name suggests, is data that our model hasn't seen before. The model has not trained on that data and will only use that data and predict a label for it.

This unseen data is called test data, a test set or a test dataset. It's used to evaluate the model's performance. Since we don't have any additional data to use as a test set, we can randomly select a small set of observations from our original dataset and set it aside. That way, we can train our model on the rest of the dataset and test it later using the test set.
Next we'll prepare our training and test datasets using scikit-learn's train_test_split function https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html. There will be two stages to this process:

- 1. The function expects the feature columns and the target column as inputs. If we look at the examples in the documentation, those two are passed to the function as X and y. As we work with the library more, we'll notice this convention often--the input data is usually stored in X and the target variable is stored in y. We'll follow the same convention here.

- 2. The function will split the dataset into a training set and a test set based on a proportion that we decide. Usually, the test set's size is about 15 to 20 percent of the dataset's. Different factors, such as the original dataset's size, can also play a part in deciding that percentage

The function splits the data randomly into each set. We'll learn why that's relevant in a future lesson.
The output of train_test_split() is a list containing 4 elements:
- The training set features.
- The test set features.
- The training set labels.
- The test set labels.

In [18]:
from sklearn.model_selection import train_test_split

x=cancer_df.drop(columns=['target'], axis=1)
y=cancer_df['target']

x_train,x_test,y_train,y_test=train_test_split(x,y, test_size=0.15, random_state=417)

In [20]:
#check the dimension of the train data sets
x_train.shape, y_train.shape

((483, 30), (483,))

In [21]:
x_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
410       11.360         17.57           72.49      399.8          0.08858   
99        14.420         19.77           94.48      642.5          0.09752   
341        9.606         16.84           61.64      280.5          0.08481   
169       14.970         16.95           96.22      685.9          0.09855   
544       13.870         20.70           89.77      584.8          0.09578   
..           ...           ...             ...        ...              ...   
78        20.180         23.97          143.70     1245.0          0.12860   
167       16.780         18.80          109.30      886.3          0.08865   
242       11.300         18.19           73.93      389.4          0.09592   
467        9.668         18.10           61.06      286.3          0.08311   
331       12.980         19.35           84.52      514.0          0.09579   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
410           0.05313         0.02783             0.021000         0.1601   
99            0.11410         0.09388             0.058390         0.1879   
341           0.09228         0.08422             0.022920         0.2036   
169           0.07885         0.02602             0.037810         0.1780   
544           0.10180         0.03688             0.023690         0.1620   
..                ...             ...                  ...            ...   
78            0.34540         0.37540             0.160400         0.2906   
167           0.09182         0.08422             0.065760         0.1893   
242           0.13250         0.15480             0.028540         0.2054   
467           0.05428         0.01479             0.005769         0.1680   
331           0.11250         0.07107             0.029500         0.1761   

     mean fractal dimension  ...  worst radius  worst texture  \
410                 0.05913  ...         13.05          36.32   
99                  0.06390  ...         16.33          30.86   
341                 0.07125  ...         10.75          23.07   
169                 0.05650  ...         16.11          23.00   
544                 0.06688  ...         15.05          24.75   
..                      ...  ...           ...            ...   
78                  0.08142  ...         23.37          31.72   
167                 0.05534  ...         20.05          26.30   
242                 0.07669  ...         12.58          27.96   
467                 0.06412  ...         11.15          24.62   
331                 0.06540  ...         14.42          21.95   

     worst perimeter  worst area  worst smoothness  worst compactness  \
410            85.07       521.3            0.1453             0.1622   
99            109.50       826.4            0.1431             0.3026   
341            71.25       353.6            0.1233             0.3416   
169           104.60       793.7            0.1216             0.1637   
544            99.17       688.6            0.1264             0.2037   
..               ...         ...               ...                ...   
78            170.30      1623.0            0.1639             0.6164   
167           130.70      1260.0            0.1168             0.2119   
242            87.16       472.9            0.1347             0.4848   
467            71.11       380.2            0.1388             0.1255   
331            99.21       634.3            0.1288             0.3253   

     worst concavity  worst concave points  worst symmetry  \
410          0.18110               0.08698          0.2973   
99           0.31940               0.15650          0.2718   
341          0.43410               0.08120          0.2982   
169          0.06648               0.08485          0.2404   
544          0.13770               0.06845          0.2249   
..               ...                   ...             ...   
78           0.76810               0.25080         

In [22]:
x_test

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
30        18.630         25.11          124.80     1088.0          0.10640   
407       12.850         21.37           82.63      514.5          0.07551   
362       12.760         18.84           81.87      496.6          0.09676   
548        9.683         19.34           61.05      285.7          0.08491   
112       14.260         19.65           97.83      629.9          0.07837   
..           ...           ...             ...        ...              ...   
33        19.270         26.47          127.90     1162.0          0.09401   
236       23.210         26.97          153.50     1670.0          0.09509   
562       15.220         30.62          103.40      716.9          0.10480   
513       14.580         13.66           94.29      658.8          0.09832   
260       20.310         27.06          132.90     1288.0          0.10000   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
30            0.18870         0.23190             0.124400         0.2183   
407           0.08316         0.06126             0.018670         0.1580   
362           0.07952         0.02688             0.017810         0.1759   
548           0.05030         0.02337             0.009615         0.1580   
112           0.22330         0.30030             0.077980         0.1704   
..                ...             ...                  ...            ...   
33            0.17190         0.16570             0.075930         0.1853   
236           0.16820         0.19500             0.123700         0.1909   
562           0.20870         0.25500             0.094290         0.2128   
513           0.08918         0.08222             0.043490         0.1739   
260           0.10880         0.15190             0.093330         0.1814   

     mean fractal dimension  ...  worst radius  worst texture  \
30                  0.06197  ...         23.15          34.01   
407                 0.06114  ...         14.40          27.01   
362                 0.06183  ...         13.75          25.99   
548                 0.06235  ...         10.93          25.59   
112                 0.07769  ...         15.30          23.73   
..                      ...  ...           ...            ...   
33                  0.06261  ...         24.15          30.90   
236                 0.06309  ...         31.01          34.51   
562                 0.07152  ...         17.52          42.79   
513                 0.05640  ...         16.76          17.24   
260                 0.05572  ...         24.33          39.16   

     worst perimeter  worst area  worst smoothness  worst compactness  \
30            160.50      1670.0           0.14910            0.42570   
407            91.63       645.8           0.09402            0.19360   
362            87.82       579.7           0.12980            0.18390   
548            69.10       364.2           0.11990            0.09546   
112           107.00       709.0           0.08949            0.41930   
..               ...         ...               ...                ...   
33            161.40      1813.0           0.15090            0.65900   
236           206.00      2944.0           0.14810            0.41260   
562           128.70       915.0           0.14170            0.79170   
513           108.50       862.0           0.12230            0.19280   
260           162.30      1844.0           0.15220            0.29450   

     worst concavity  worst concave points  worst symmetry  \
30            0.6133               0.18480          0.3444   
407           0.1838               0.05601          0.2488   
362           0.1255               0.08312          0.2744   
548           0.0935               0.03846          0.2552   
112           0.6783               0.15050          0.2398   
..               ...                   ...             ...   
33            0.6091               0.17850         

In [23]:
y_train

410    1
99     0
341    1
169    1
544    1
      ..
78     0
167    0
242    1
467    1
331    1
Name: target, Length: 483, dtype: int32

In [24]:
y_test

30     0
407    1
362    1
548    1
112    1
      ..
33     0
236    0
562    0
513    1
260    0
Name: target, Length: 86, dtype: int32

##  Building and Training a Classifier
We have our training data ready! We can now train a model:

<img src='train.svg' width=400 height=400>

scikit-learn offers many machine learning models https://scikit-learn.org/stable/supervised_learning.html. We can pick any one of the classification models and use it out of the box.
That's exactly what we'll do. We'll use the Linear Support Vector Classification model https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html. We can check the documentation to see what kind of input parameters it can take. scikit-learn makes it easy for us to use the model without needing to understand the minute details of the algorithm. We'll observe this soon.

- There are only two steps for building and training the model:

- 1 We first instantiate the model. This is similar to how we would instantiate a Python class.

- 2 This step doesn't take any training data as input. We can, however, define and set the values for the parameters we saw in the documentation linked above.

- We fit the model onto the training data features (X) and labels (y). Fitting the model is the same as training our model.

In [29]:
from sklearn.svm import LinearSVC

#instantiate the model
model=LinearSVC(penalty='l2', loss='squared_hinge', C=10, random_state=417)

model.fit(x_train, y_train)

y_pred=model.predict(x_test)
y_pred[0:10]

C:\Users\512GB\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


array([0, 0, 1, 1, 1, 0, 0, 1, 0, 1])

## Evaluating the Model on Test Set I
We have successfully trained our model! But we don't know how well it performs.

How can we measure the performance of a model?

<img src='evaluate.svg' width=400 height=400>

One of the most common ways to evaluate a classifier is to look at how accurate its predictions are. For a supervised learning task, we already know which class a particular observation belongs to.

We can use our model to predict the labels of our test data. We can then calculate the accuracy of our model by comparing those predictions to the actual labels. But we don't need to do that all by ourselves.
scikit-learn provides a method score()https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html#sklearn.svm.LinearSVC.score that can calculate that accuracy

In [34]:
from sklearn.metrics import classification_report,confusion_matrix

train_accuracy=model.score(x_train, y_train)
test_accuracy=model.score(x_test, y_test)

print('Train Accuracy is :', train_accuracy)
print('Test Accuracy is:',  test_accuracy)

Train Accuracy is : 0.8509316770186336
Test Accuracy is: 0.8255813953488372


##  Fine-tuning the Model
Our model got an accuracy score of approximately 82.56. That means roughly 82.56% of our predictions were correct.

That's a pretty good result, but building machine learning models is an iterative process. We can try to improve upon our result.
<img src='finetune.svg' width=400 height=400>

When we instantiated the LinearSVC model, we defined the following parameters:

- penalty="l2"
- loss="squared_hinge"
- C=10
While we don't know what these values imply for our model, we can experiment with them.
- In an attempt to improve upon our result, we will:

Change the values to one of the above parameters.
Add and set another parameter from the documentation https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html#sklearn-svm-linearsvc.

In [37]:
#fine tune the model
model=LinearSVC(penalty='l2' ,loss='squared_hinge' ,C=20, max_iter=3500, random_state=417)

#instantiate the model
model.fit(x_train, y_train)
pred=model.predict(x_test)

C:\Users\512GB\anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


## Evaluating the Model on Test Set II
In the last screen, we fine-tuned our model. We don't know yet if that improved our model's performance.

Before we evaluate our model on the test set, let's look at the following plot:

<img src='plot.svg' width=400 height=400>

We can notice that there are yellow dots on the left side and there are purple on the right side of the decision boundary. The above is meant to be a simplified representation of how the decision boundary can't always capture the complexity of the data.

What happens when we add a point to the above plot? Will it be correctly classified?

We can't be sure. Our model is not going to be perfect. It's not going to give us a 100% accuracy, even if we keep experimenting with different parameter values.

Let's see how our updated model performs on the test set

In [38]:
train_accuracy=model.score(x_train, y_train)
test_accuracy=model.score(x_test, y_test)

print('Train Accuracy is :', train_accuracy)
print('Test Accuracy is:',  test_accuracy)

Train Accuracy is : 0.9006211180124224
Test Accuracy is: 0.872093023255814


## Experimentation vs Fundamentals
Our accuracy increased!

By just playing around with a couple of parameters, we were able to improve how our model performed. Our model is now better at predicting whether a patient has breast cancer.

It's really amazing how we managed to:

- Train a randomly selected model from scikit-learn.
- Experiment with a few of its parameters, and
- Get it to predict, with reasonably high accuracy, whether a patient has breast cancer.
We managed to do that without understanding:

- What the model actually does.
- What the different parameters of that model are for.
- What most of the dataset's features are.
While there is clearly a case to be made for the value of being able to quickly iterate and experiment with different models, we can't lose sight of the bigger picture either.

Remember we set max_iter to 3500? If we'd set it to 3000 instead, our accuracy would have dropped precipitously!

What would we have done in that situation? There are 12 parameters in scikit-learn's LinearSVC.

- How many of those parameters could we experiment with?
- What possible permutations and combinations of values could we have tried?
- What happens if we lower C instead of increasing it? 
- What does C actually do in the model? Would lowering it make our model perform better on the test set?
- If we were to get more data with some new features, would our selected values yield similar results or would we have to go through the process all over again?
- What if we had millions of observations instead? Could we afford to run any model any number of times on such a large dataset? Which model would even be suitable for such a large dataset?

Experimentation and iteration through direct application have their place in the field of machine learning. However, trying to achieve good-to-great results given certain constraints is a challenging task. Blind experimentation won't get us far.

Understanding the fundamentals or understanding how a machine learning algorithm works "under the hood" opens up additional insights to rely on when trying to answer some of the questions above. It can allow us to experiment and iterate from an informed perspective.

In the next few lessons, we'll learn from that other perspective. We'll learn about a different machine learning algorithm, implement it from scratch, and then use scikit-learn again.